### ISSUE TITLE
[Feature Request] Add json logging support

### ISSUE THREAD

**Issue by shachar-ash** [rfjakob/earlyoom#218]
earlyoom current logging is pretty hard to parse, adding a JSON logging support will make our life much easier.

Thanks!

**Comment by ofirnn** [comment]
That's really needed and helpful!

**Comment by rfjakob** [comment]
Hmm, what kind of information would you like to see in the json? Can you post example json messages?

**Comment by hakavlad** [comment]
journald provides json output:
```
sudo journalctl --output=json -u earlyoom.service
```

```
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=62319a;b=3957642b91db46aeb6ea566e41d1e893;m=1489e1251;t=5ab86c23e5e9b;x=ff6b31d933abaf6a", "__REALTIME_TIMESTAMP" : "1595970156387995", "__MONOTONIC_TIMESTAMP" : "5513286225", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "_SYSTEMD_SLICE" : "system.slice", "_TRANSPORT" : "stdout", "SYSLOG_IDENTIFIER" : "earlyoom", "_PID" : "7758", "_UID" : "61876", "_GID" : "61876", "_COMM" : "earlyoom", "_EXE" : "/usr/local/bin/earlyoom", "_CMDLINE" : "/usr/local/bin/earlyoom -r 3600 -m 4", "_CAP_EFFECTIVE" : "4020", "_SYSTEMD_CGROUP" : "/system.slice/earlyoom.service", "_SYSTEMD_UNIT" : "earlyoom.service", "_SYSTEMD_INVOCATION_ID" : "d511373d58ac47c5b02fcce0d00a3c7c", "MESSAGE" : "low memory! at or below SIGTERM limits: mem  4.00%, swap 10.00%" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=62319b;b=3957642b91db46aeb6ea566e41d1e893;m=148a02268;t=5ab86c2406eb3;x=6a7802014e311f28", "__REALTIME_TIMESTAMP" : "1595970156523187", "__MONOTONIC_TIMESTAMP" : "5513421416", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "_SYSTEMD_SLICE" : "system.slice", "_TRANSPORT" : "stdout", "SYSLOG_IDENTIFIER" : "earlyoom", "_PID" : "7758", "_UID" : "61876", "_GID" : "61876", "_COMM" : "earlyoom", "_EXE" : "/usr/local/bin/earlyoom", "_CMDLINE" : "/usr/local/bin/earlyoom -r 3600 -m 4", "_CAP_EFFECTIVE" : "4020", "_SYSTEMD_CGROUP" : "/system.slice/earlyoom.service", "_SYSTEMD_UNIT" : "earlyoom.service", "_SYSTEMD_INVOCATION_ID" : "d511373d58ac47c5b02fcce0d00a3c7c", "MESSAGE" : "sending SIGTERM to process 7760 uid 1000 \"tail\": badness 869, VmRSS 8226 MiB" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=62319c;b=3957642b91db46aeb6ea566e41d1e893;m=148c1467d;t=5ab86c26192c8;x=a018b7242cc5c3e1", "__REALTIME_TIMESTAMP" : "1595970158695112", "__MONOTONIC_TIMESTAMP" : "5515593341", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "_SYSTEMD_SLICE" : "system.slice", "_TRANSPORT" : "stdout", "SYSLOG_IDENTIFIER" : "earlyoom", "_PID" : "7758", "_UID" : "61876", "_GID" : "61876", "_COMM" : "earlyoom", "_EXE" : "/usr/local/bin/earlyoom", "_CMDLINE" : "/usr/local/bin/earlyoom -r 3600 -m 4", "_CAP_EFFECTIVE" : "4020", "_SYSTEMD_CGROUP" : "/system.slice/earlyoom.service", "_SYSTEMD_UNIT" : "earlyoom.service", "_SYSTEMD_INVOCATION_ID" : "d511373d58ac47c5b02fcce0d00a3c7c", "MESSAGE" : "process exited after 2.3 seconds" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=6231a3;b=3957642b91db46aeb6ea566e41d1e893;m=14d4c2452;t=5ab86c6ec709c;x=d8610066a54f295c", "__REALTIME_TIMESTAMP" : "1595970234904732", "__MONOTONIC_TIMESTAMP" : "5591802962", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "_SYSTEMD_SLICE" : "system.slice", "_TRANSPORT" : "stdout", "SYSLOG_IDENTIFIER" : "earlyoom", "_PID" : "7758", "_UID" : "61876", "_GID" : "61876", "_COMM" : "earlyoom", "_EXE" : "/usr/local/bin/earlyoom", "_CMDLINE" : "/usr/local/bin/earlyoom -r 3600 -m 4", "_CAP_EFFECTIVE" : "4020", "_SYSTEMD_CGROUP" : "/system.slice/earlyoom.service", "_SYSTEMD_UNIT" : "earlyoom.service", "_SYSTEMD_INVOCATION_ID" : "d511373d58ac47c5b02fcce0d00a3c7c", "MESSAGE" : "mem avail:   351 of  9823 MiB ( 3.58%), swap free: 1909 of 19647 MiB ( 9.72%)" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=6231a4;b=3957642b91db46aeb6ea566e41d1e893;m=14d4c2452;t=5ab86c6ec709c;x=ff6b31d933abaf6a", "__REALTIME_TIMESTAMP" : "1595970234904732", "__MONOTONIC_TIMESTAMP" : "5591802962", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "_SYSTEMD_SLICE" : "system.slice", "_TRANSPORT" : "stdout", "SYSLOG_IDENTIFIER" : "earlyoom", "_PID" : "7758", "_UID" : "61876", "_GID" : "61876", "_COMM" : "earlyoom", "_EXE" : "/usr/local/bin/earlyoom", "_CMDLINE" : "/usr/local/bin/earlyoom -r 3600 -m 4", "_CAP_EFFECTIVE" : "4020", "_SYSTEMD_CGROUP" : "/system.slice/earlyoom.service", "_SYSTEMD_UNIT" : "earlyoom.service", "_SYSTEMD_INVOCATION_ID" : "d511373d58ac47c5b02fcce0d00a3c7c", "MESSAGE" : "low memory! at or below SIGTERM limits: mem  4.00%, swap 10.00%" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=6231a5;b=3957642b91db46aeb6ea566e41d1e893;m=14d4d0346;t=5ab86c6ed4f91;x=4e7e2f6a80efe2ea", "__REALTIME_TIMESTAMP" : "1595970234961809", "__MONOTONIC_TIMESTAMP" : "5591860038", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "_SYSTEMD_SLICE" : "system.slice", "_TRANSPORT" : "stdout", "SYSLOG_IDENTIFIER" : "earlyoom", "_PID" : "7758", "_UID" : "61876", "_GID" : "61876", "_COMM" : "earlyoom", "_EXE" : "/usr/local/bin/earlyoom", "_CMDLINE" : "/usr/local/bin/earlyoom -r 3600 -m 4", "_CAP_EFFECTIVE" : "4020", "_SYSTEMD_CGROUP" : "/system.slice/earlyoom.service", "_SYSTEMD_UNIT" : "earlyoom.service", "_SYSTEMD_INVOCATION_ID" : "d511373d58ac47c5b02fcce0d00a3c7c", "MESSAGE" : "sending SIGTERM to process 7767 uid 1000 \"python3\": badness 874, VmRSS 8401 MiB" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=6231a6;b=3957642b91db46aeb6ea566e41d1e893;m=14d5e8876;t=5ab86c6fed4bf;x=ca7e5ae0f23bc66d", "__REALTIME_TIMESTAMP" : "1595970236110015", "__MONOTONIC_TIMESTAMP" : "5593008246", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "_SYSTEMD_SLICE" : "system.slice", "_TRANSPORT" : "stdout", "SYSLOG_IDENTIFIER" : "earlyoom", "_PID" : "7758", "_UID" : "61876", "_GID" : "61876", "_COMM" : "earlyoom", "_EXE" : "/usr/local/bin/earlyoom", "_CMDLINE" : "/usr/local/bin/earlyoom -r 3600 -m 4", "_CAP_EFFECTIVE" : "4020", "_SYSTEMD_CGROUP" : "/system.slice/earlyoom.service", "_SYSTEMD_UNIT" : "earlyoom.service", "_SYSTEMD_INVOCATION_ID" : "d511373d58ac47c5b02fcce0d00a3c7c", "MESSAGE" : "process exited after 1.2 seconds" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=6231f8;b=3957642b91db46aeb6ea566e41d1e893;m=15141ecb7;t=5ab86cae23902;x=a2b8a7bb7fee0ed5", "__REALTIME_TIMESTAMP" : "1595970301344002", "__MONOTONIC_TIMESTAMP" : "5658242231", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "SYSLOG_IDENTIFIER" : "systemd", "_UID" : "0", "_GID" : "0", "_TRANSPORT" : "journal", "_PID" : "1", "_COMM" : "systemd", "_EXE" : "/lib/systemd/systemd", "_CMDLINE" : "/sbin/init", "_CAP_EFFECTIVE" : "3fffffffff", "_SYSTEMD_CGROUP" : "/init.scope", "_SYSTEMD_UNIT" : "init.scope", "_SYSTEMD_SLICE" : "-.slice", "CODE_FILE" : "../src/core/unit.c", "CODE_LINE" : "1493", "CODE_FUNCTION" : "unit_status_log_starting_stopping_reloading", "MESSAGE_ID" : "de5b426a63be47a7b6ac3eaac82e2f6f", "UNIT" : "earlyoom.service", "MESSAGE" : "Stopping Early OOM Daemon...", "_SOURCE_REALTIME_TIMESTAMP" : "1595970294400759" }
{ "__CURSOR" : "s=ccf7523668294683b810e2ec1ec53163;i=62320a;b=3957642b91db46aeb6ea566e41d1e893;m=151425318;t=5ab86cae29f64;x=a004faa5115c967c", "__REALTIME_TIMESTAMP" : "1595970301370212", "__MONOTONIC_TIMESTAMP" : "5658268440", "_BOOT_ID" : "3957642b91db46aeb6ea566e41d1e893", "PRIORITY" : "6", "_MACHINE_ID" : "9d230526b63a40f5b6ff08c3cae75f08", "_HOSTNAME" : "PC", "SYSLOG_FACILITY" : "3", "SYSLOG_IDENTIFIER" : "systemd", "_UID" : "0", "_GID" : "0", "CODE_FILE" : "../src/core/job.c", "CODE_LINE" : "805", "CODE_FUNCTION" : "job_log_status_message", "RESULT" : "done", "_TRANSPORT" : "journal", "_PID" : "1", "_COMM" : "systemd", "_EXE" : "/lib/systemd/systemd", "_CMDLINE" : "/sbin/init", "_CAP_EFFECTIVE" : "3fffffffff", "_SYSTEMD_CGROUP" : "/init.scope", "_SYSTEMD_UNIT" : "init.scope", "_SYSTEMD_SLICE" : "-.slice", "MESSAGE_ID" : "9d1aaa27d60140bd96365438aad20286", "UNIT" : "earlyoom.service", "MESSAGE" : "Stopped Early OOM Daemon.", "_SOURCE_REALTIME_TIMESTAMP" : "1595970295062858" }
```

**Comment by shachar-ash** [comment]
> Hmm, what kind of information would you like to see in the json? Can you post example json messages?

The same information currently displayed by Earlyoom, maybe just add a timestamp to it. 

> {"timestamp": "2014-05-24T22:47:48", "mem_avail": "9248", "mem_avail_pct": "26", "swap_avail": "100", "swap_avail_pct": "14", "pid_killed":"123", "process_cmdline": "/bin/bash /home/sergio/foo.sh"}

**Comment by rfjakob** [comment]
I'm afraid I don't want to get into the complexities of generating JSON from C.

For monitoring free memory there are a lot of other tools out there, and for noticing when earlyoom kills something, you can use the D-Bus notifications!
